Source/Core/VideoCommon/UberShaderPixel.cpp

```diff
@@ -0,0 +1,687 @@
+
+	if (!g_ActiveConfig.backend_info.bSupportsBitfield)
+	{
+		out.Write(
+			"uint bitfieldExtract(uint val, int off, int size) {\n"
+			"	// This built-in function is only support in OpenGL 4.0+ and ES 3.1+\n"
+			"	// Microsoft's HLSL compiler automatically optimises this to a bitfield extract instruction.\n"
+			"	uint mask = uint((1 << size) - 1);\n"
+			"	return uint(val >> off) & mask;\n"
+			"}\n\n");
+	}
+
+	// =====================
+	//   Texture Sampling
+	// =====================
+
+	if (g_ActiveConfig.backend_info.bSupportsDynamicSamplerIndexing)
+	{
+		// Doesn't look like directx supports this. Oh well the code path is here just incase it supports this in the future.
+		out.Write(
+			"int4 sampleTexture(uint sampler_num, float2 uv) {\n");
+		if (ApiType == API_OPENGL) out.Write(
+			"	return int4(texture(samp[sampler_num], float3(uv, 0.0)) * 255.0);\n");
+		else if (ApiType == API_D3D) out.Write(
+			"	return int4(Tex[sampler_num].Sample(samp[sampler_num], float3(uv, 0.0)) * 255.0);\n");
+		out.Write(
+			"}\n\n");
+	}
+	else
+	{
+		out.Write(
+			"int4 sampleTexture(uint sampler_num, float2 uv) {\n"
+			"	// This is messy, but DirectX, OpenGl 3.3 and Opengl ES 3.0 doesn't support dynamic indexing of the sampler array\n"
+			"	// With any luck the shader compiler will optimise this if the hardware supports dynamic indexing.\n"
+			"	switch(sampler_num) {\n");
+		for (int i = 0; i < 8; i++)
+		{
+			if (ApiType == API_OPENGL) out.Write(
+				"	case %du: return int4(texture(samp[%d], float3(uv, 0.0)) * 255.0);\n", i, i);
+			else if (ApiType == API_D3D) out.Write(
+				"	case %du: return int4(Tex[%d].Sample(samp[%d], float3(uv, 0.0)) * 255.0);\n", i, i, i);
+		}
+		out.Write(
+			"	}\n"
+			"}\n\n");
+	}
+
+	// ======================
+	//   TEV's Special Lerp
+	// ======================
+
+	out.Write(
+		"// One channel worth of TEV's Linear Interpolate, plus bias, add/subtract and scale\n"
+		"int tevLerp(int A, int B, int C, int D, uint bias, bool op, uint shift) {\n"
+		"	C += C >> 7; // Scale C from 0..255 to 0..256\n"
+		"	int lerp = (A << 8) + (B - A)*C;\n"
+		"	if (shift != 3u) {\n"
+		"		lerp = lerp << shift;\n"
+		"		lerp = lerp + (op ? 127 : 128);\n"
+		"	}\n"
+		"	int result = lerp >> 8;\n"
+		"\n"
+		"	// Add/Subtract D (and bias)\n"
+		"	if (bias == 1u) result += 128;\n"
+		"	else if (bias == 2u) result -= 128;\n"
+		"	if(op) // Subtract\n"
+		"		result = D - result;\n"
+		"	else // Add\n"
+		"		result = D + result;\n"
+		"\n"
+		"	// Most of the Shift was moved inside the lerp for improved percision\n"
+		"	// But we still do the divide by 2 here\n"
+		"	if (shift == 3u)\n"
+		"		result = result >> 1;\n"
+		"	return result;\n"
+		"}\n\n");
+
+	// =======================
+	//   TEV's Color Compare
+	// =======================
+
+	out.Write(
+		"// Implements operations 0-5 of tev's compare mode,\n"
+		"// which are common to both color and alpha channels\n"
+		"bool tevCompare(uint op, int3 color_A, int3 color_B) {\n"
```

Comment:
Do you think it's worth to return a bool4 instead? So also case 6u and 7u could be handled.